### DISCUSSION TITLE
`int` being a subtype of `float` and type narrowing

### DISCUSSION THREAD

**Discussion by decorator-factory** [python/typing#1209] [General]
Suppose that `x` has type `str | float`. Should `if not isinstance(x, float):` narrow `x` to `str`, or to `str | int`? If it narrows it to `str`, it seems like the checking will be "unsound", because `int`s are considered assignable to `float`, but the `isinstance` check doesn't consider integers to be floats.

Right now both [mypy](https://mypy-play.net/?mypy=latest&python=3.10&gist=9d81e6699a5be945d9dc530345125370) and [pyright](https://pyright-playground.decorator-factory.su/?gzip=H4sIAMxhomIC_0tJTVNI06iwUkjLyU8sUahRKC4p0lTQtVPwy89LteJSAILMNIXM4sy84pLEvORUjQodiFJNiCQIFBRl5pVoVChoKxjrGZpogsVTc4pTsalQUlRUVNLk4qpUsAXaa2KkyVWUWpaamBNfUlmQqlGpyQUAzr_MXpEAAAA%3D) think this is OK:
```py
def f(x: float | str) -> None:
    if isinstance(x, float):
        print(x + 3.14)
    else:
        print(x + "!!!")
```
But a call like `f(42)` (also considered legal) will cause a runtime error, because the execution will go to the `else` branch.

I don't have any real-world examples for this issue, so this is purely hypothetical for now.

**Comment by erictraut** [comment]
Yes, `isinstance` narrowing is inexact. You can contrive many edge cases where it produces behavior that's not strictly type safe. 

Consider that a class could be a subclass of both `str` and `float` at the same time, yet it's useful for the `isinstance` narrowing logic to assume that it is one or the other. Also, classes can implement a `__instancecheck__` that "lies" about whether two classes have a subclass relationship. And there are cases (like the one you mention above) where the type system indicates that there is a subclass relationship when such a relationship doesn't exist at runtime.

So yes, there are plenty of holes here. This is one of those situations where practicality trumps purity.

**Reply by Akuli** [reply]
I can't think of a practical use for floats where `int`s wouldn't be acceptable too. So to me, it seems that `isinstance(x, float)` is buggy, unless you also check for `isinstance(x, int)`.

Maybe it would be better to narrow `x` to `int | str` in the `else`? I don't think that would cause much false positives.
